### PR TITLE
[AP-851] Fix for config json files

### PR DESCRIPTION
**Pull request by amofakhar**
## Context

https://transferwise.atlassian.net/browse/AP-851

## Problem

In some cases when json config files are corrupted or other services are writing on them, if we run the taps they would raise exception because of invalid json file.

## Proposed changes

By this fix we will validate the json config files before running the tap and if it is an invalid file then retires for some specific times to be sure file is fixed.
https://transferwise.atlassian.net/browse/AP-851

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [x] CI checks pass with my changes
- [x] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [x] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: tests/units/cli/test_commands.py [hunk 1]
```diff
@@ -662,3 +678,189 @@ def test_run_command(self):
...
+            )
+
+    def test_tap_config_valid_if_json_property_is_none(self):
+        """Test TapConfig is valid if a json property is None"""
+        with TemporaryDirectory() as temp_dir:
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            for case_number in range(3):
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:800]: same comment here about cases

File: tests/units/cli/test_commands.py [hunk 2]
```diff
@@ -662,3 +678,189 @@ def test_run_command(self):
...
+        with TemporaryDirectory() as temp_dir:
+            invalid_json_file = f'{temp_dir}/invalid_file.json'
+            valid_json_file = f'{temp_dir}/valid_file.json'
+
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            for invalid_content in invalid_file_contents:
+                with open(invalid_json_file, 'w', encoding='utf-8') as invalid_file:
+                    invalid_file.write(invalid_content)
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:820]: same comment about using writelines

File: tests/units/cli/test_commands.py [hunk 3]
```diff
@@ -662,3 +678,189 @@ def test_run_command(self):
             )
         assert os.path.isfile('test.log.failed')
         os.remove('test.log.failed')
+
+
+class TestConfigValidation:
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:683]: I see there is a lot of repetition of creating valid and invalid files, maybe it's better to create one valid, one invalid and 1 empty file and reuse them throughout the test suite.

File: pipelinewise/cli/commands.py [hunk 4]
```diff
@@ -29,13 +35,84 @@
 )
 
 
-class RunCommandException(Exception):
+def is_invalid_json_file(json_file_path, file_property):
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py:38]: - [ ] Refactor the method

File: pipelinewise/cli/commands.py [hunk 5]
```diff
@@ -29,13 +35,84 @@
...
     """
-    Custom exception to raise when run command fails
+    checking if input file is a valid json or not, in some cases it is allowed to have an empty file, or it is allowed
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py:40]: - [ ] Fix doctoring

File: pipelinewise/cli/commands.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py]: It should be an internal method

File: pipelinewise/cli/commands.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py]: It is an internal method now

File: pipelinewise/cli/commands.py [hunk 6]
```diff
@@ -29,13 +35,84 @@
...
+            invalid_json = True
+            time.sleep(PARAMS_VALIDATION_RETRY_PERIOD_SEC)
+        else:
+            invalid_json = False
+            break
+    return invalid_json
+
+
+# pylint: disable=function-redefined)
+class TapParams(TapParams):
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py]: - [ ] fix it

File: pipelinewise/cli/commands.py [hunk 7]
```diff
@@ -29,13 +35,84 @@
...
+        list_of_params_in_json_file = {
+            'config': {'file_must_exists': True, 'allowed_empty': False},
+            'properties': {'file_must_exists': True, 'allowed_empty': False},
+            'state': {'file_must_exists': False, 'allowed_empty': True}
+        }
+        for param, file_property in list_of_params_in_json_file.items():
+            invalid_json = do_json_conf_validation(
+                json_file=kwargs[param],
+                file_property=file_property
+               ) if kwargs.get(param) else False
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py]: - [ ] fix it

File: pipelinewise/cli/commands.py [hunk 8]
```diff
@@ -29,13 +35,84 @@
...
+class TargetParams(TargetParams):
+    """
+    for overriding TargetParam init and validating json properties
+    """
+    # pylint: disable=unused-argument
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        invalid_json = do_json_conf_validation(
+            json_file=kwargs.get('config'),
+            file_property={'file_must_exists': True, 'allowed_empty': False}) if kwargs.get('config') else False
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply pipelinewise/cli/commands.py]: - [ ] fix it

File: tests/units/cli/test_commands.py [hunk 9]
```diff
@@ -662,3 +678,189 @@ def test_run_command(self):
...
+    def test_tap_config_json_validation(self):
+        """Test it retries if any json file is invalid"""
+        with TemporaryDirectory() as temp_dir:
+
+            test_json_file = f'{temp_dir}/test_file.json'
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            for case_number in range(3):
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] refactor it

File: tests/units/cli/test_commands.py [hunk 10]
```diff
@@ -662,3 +678,189 @@ def test_run_command(self):
...
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            empty_file = f'{temp_dir}/empty_stat.json'
+
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            # 1. asserting if raises exception on invalid contents
+            for invalid_content in invalid_file_contents:
+                with open(invalid_json_file, 'w', encoding='utf-8') as invalid_file:
+                    invalid_file.write(invalid_content)
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py:782]: It is not same file, they are multiple files with different contents

File: tests/units/cli/test_commands.py (hunk 2 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py:828]: answered above

File: tests/units/cli/test_commands.py (hunk 1 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] refactor it

File: tests/units/cli/test_commands.py (hunk 3 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py:683]: This helps each test be isolated and using a new file and also in some scenarios an invalid file will be fixed and become a valid one.

File: tests/units/cli/test_commands.py [hunk 11]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+            valid_json_file = f'{temp_dir}/valid_file.json'
+
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            for invalid_content in invalid_file_contents:
+                with open(invalid_json_file, 'w', encoding='utf-8') as invalid_file:
+                    invalid_file.write(invalid_content)
+
+                for case_number in range(2):
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:835]: same old comment about cases and for loops

File: tests/units/cli/test_commands.py [hunk 12]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+
+    def test_tap_config_valid_if_json_property_is_none(self):
+        """Test TapConfig is valid if a json property is None"""
+        with TemporaryDirectory() as temp_dir:
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            test_cases = (
+                (None, valid_json_file, valid_json_file),
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:809]: test case should fail, `config` is a mandatory parameter to pass to the tap in both discovery and sync modes

File: tests/units/cli/test_commands.py [hunk 13]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+            )
+
+    def test_tap_config_valid_if_json_property_is_none(self):
+        """Test TapConfig is valid if a json property is None"""
+        with TemporaryDirectory() as temp_dir:
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            test_cases = (
```
Comment (Samira-El) [inline tests/units/cli/test_commands.py:808]: please break this into separate test cases(methods)

File: tests/units/cli/test_commands.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py:782]: They are not different scenarios,! it is just same test with different file contents (invalid contents). no need to have separate test cases.

File: tests/units/cli/test_commands.py [hunk 14]
```diff
@@ -12,6 +18,12 @@ class TestCommands:
     Unit tests for PipelineWise CLI commands functions
     """
 
+    # pylint: disable=redefined-outer-name
+    @pytest.fixture(autouse=True)
+    def mock_json_validation(self, mocker):  # noqa: F811
+        """we are using some config files which does not exist, so we patch the method that verifies the json files"""
+        mocker.patch('pipelinewise.cli.commands._verify_json_file', return_value=True)
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py:25]: By looking at the other tests I thought using Pytest is the goal for this project - which is not correct!
I will fix it.

- [ ] use unittest mock

File: tests/units/cli/test_commands.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py:691]: - [ ] Move to a separate file

File: tests/units/cli/test_commands.py [hunk 15]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+
+    def test_tap_config_json_validation(self):
+        """Test it retries if any json file is invalid"""
+        test_cases = (
+            {'config': 'invalid', 'properties': 'valid', 'state': 'valid'},
+            {'config': 'valid', 'properties': 'invalid', 'state': 'valid'},
+            {'config': 'valid', 'properties': 'valid', 'state': 'invalid'},
+        )
+        for json_files_situation in test_cases:
+            self._assert_retry_validation_of_json_file(json_files_situation)
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: I prefer not if the goal is not using pytest!

File: tests/units/cli/test_commands.py [hunk 16]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+        with TemporaryDirectory() as temp_dir:
+            valid_json_file = f'{temp_dir}/valid_file.json'
+            not_exists_state_file = f'{temp_dir}/not_exists.json'
+            with open(valid_json_file, 'w', encoding='utf-8') as valid_file:
+                json.dump({'foo': 'bar'}, valid_file)
+
+            self._assert_tap_config(config=valid_json_file, properties=valid_json_file, state=not_exists_state_file)
+
+    # pylint: disable=invalid-name
+    def test_tap_config_json_valid_if_state_file_is_empty_and_raise_exception_on_invalid_content(self):
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] make separate test cases

File: tests/units/cli/test_commands.py (hunk 11 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] Fix it

File: tests/units/cli/test_commands.py [hunk 17]
```diff
@@ -662,3 +678,202 @@ def test_run_command(self):
...
+            # It starts with an invalid file and since the main method is retrying we fix the file after some seconds
+            fixed_file = self.AsyncWriteJsonFile(test_json_file, '{"foo": "bar"}', self.sec_to_repair_json_file)
+            fixed_file.start()
+            fixed_file.join()
+
+            self._assert_target_config(config=test_json_file)
+
+    def test_target_config_valid_if_json_property_is_none(self):
+        """Test TargetConfig is valid if config is None"""
+        self._assert_target_config(config=None)
```
[the remark this answers is not in the document]
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] Fix it

File: tests/units/cli/test_commands.py (hunk 12 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] Fix it

File: tests/units/cli/test_commands.py (hunk 13 above)
Comment (amofakhar) [reply tests/units/cli/test_commands.py]: - [ ] Make separate test cases

File: pipelinewise/cli/commands.py [hunk 18]
```diff
@@ -16,26 +19,104 @@
...
 )
-TargetParams = namedtuple('TargetParams', ['id', 'type', 'bin', 'python_bin', 'config'])
+OriginalTargetParams = namedtuple('TargetParams', ['id', 'type', 'bin', 'python_bin', 'config'])
 TransformParams = namedtuple(
     'TransformParams', ['bin', 'python_bin', 'config', 'tap_id', 'target_id']
 )
 
 
-class RunCommandException(Exception):
+def _verify_json_file(json_file_path, file_must_exists, allowed_empty):
```
Comment (jeet-parekh-wise) [inline pipelinewise/cli/commands.py:38]: please add type hints

File: pipelinewise/cli/commands.py [hunk 19]
```diff
@@ -16,26 +19,104 @@
...
+            json.load(json_file)
+    except FileNotFoundError:
+        return not file_must_exists
+    except json.decoder.JSONDecodeError:
+        if not allowed_empty or os.stat(json_file_path).st_size != 0:
+            return False
+    return True
+
+
+def do_json_conf_validation(json_file, file_property):
```
Comment (jeet-parekh-wise) [inline pipelinewise/cli/commands.py:53]: please add type hints

File: tests/units/cli/test_config_validation.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+        with open(self.empty_file, 'w', encoding='utf-8') as empty_file:
+            empty_file.write('')
+
+    def tearDown(self):
+        self.temp_dir.cleanup()
+
+    class AsyncWriteJsonFile(threading.Thread):
+        """Helper class for asynchronous writing on a file"""
+        def __init__(self, file_name, content, waiting_time):
```
Comment (jeet-parekh-wise) [inline tests/units/cli/test_config_validation.py:33]: please add type hints

File: pipelinewise/cli/commands.py (hunk 18 above)
Comment (amofakhar) [reply pipelinewise/cli/commands.py:38]: - [ ] Fix it

File: pipelinewise/cli/commands.py (hunk 19 above)
Comment (amofakhar) [reply pipelinewise/cli/commands.py:53]: - [ ] fix it!

File: tests/units/cli/test_config_validation.py (hunk 20 above)
Comment (amofakhar) [reply tests/units/cli/test_config_validation.py:33]: - [ ] Fix it

**Review by jeet-parekh-wise — APPROVED** [review]

**Pull request merged**